### PR TITLE
Add highlight test and change scrolling

### PR DESCRIPTION
**Pull request by csillag**
This PR is an updated version of of https://github.com/hypothesis/h/pull/1950.

Adding a minimal test suite here for the TextHighlight implementation.

Also changing the scrolling API according to @tilgovi's suggestions, 
in order to avoid any false appearance that we are implementing
the proposed scrollIntoView() API [1], while we clearly don't.

[1] http://dev.w3.org/csswg/cssom-view/#dom-element-scrollintoview

### CODE REVIEW

**Comment by csillag** [comment]
Any further comments on this one?

File: tests/js/plugin/enhancedanchoring-test.coffee [hunk 1]
```diff
@@ -673,11 +676,11 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
             renderPage am.document, 9
             renderPage am.document, 10
 
-        # Now we trigger the actual action
-        anchor.scrollIntoView().then ->
+        anchor.scrollToView()
+        watchForScroll().then ->
```
Comment (tilgovi) [inline tests/js/plugin/enhancedanchoring-test.coffee]: This seems wrong. If you're calling `scrollToView` before `watchForScroll` then I think `pendingTest` hasn't been initialized yet for this run. Furthermore, if you're resolving the promise instantaneously in your spy there's no reason to do this in a callback. I'm not sure what's getting tested here.

File: tests/js/plugin/enhancedanchoring-test.coffee [hunk 2]
```diff
@@ -688,5 +691,6 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
             renderPage am.document, 9
             renderPage am.document, 10
 
-        anchor.scrollIntoView().then ->
-          assert.called anchor.highlight[10].scrollIntoView
+        anchor.scrollToView()
+        watchForScroll().then ->
```
Comment (tilgovi) [inline tests/js/plugin/enhancedanchoring-test.coffee]: Same here. I'm not following the flow.

File: tests/js/plugin/enhancedanchoring-test.coffee (hunk 1 above)
Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > If you're calling `scrollToView` before `watchForScroll` then I think `pendingTest` hasn't been initialized yet for this run. 

This is the sequence of events happening:
 * We call the Anchor's `scrollToView()`
 * It triggers a page rendering, but it will only happen a bit later.
 * Then we call `watchForScroll()`, which sets `pendingTest`.
 * When the (fake) page is rendered, the fake Highlight is created, and
 * `scrollToView()` on the Highlight is called,
 * which resolves the promise (using the functions stored in `pendingTest`)
 * and then we run our asserts.
 
> Furthermore, if you're resolving the promise instantaneously in your spy there's no reason to do this in a callback.

The spy is for Highlight's `scrollToView()`. That's instantaneous. However, here we are testing the Anchor's `scrollToView()`, which does *not* complete it's task instantaneously, since it needs to wait for the asynchronous page rendering to happen.

>  I'm not sure what's getting tested here.

The two-phased scrolling behavior, which includes starting scrolling (by triggering a page rendering), then waiting for it to finish, and then finishing the scrolling (by calling `scrollToView`() on the right highlight.).

File: tests/js/plugin/enhancedanchoring-test.coffee (hunk 2 above)
Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > Same here. I'm not following the flow.

See my explanation above. If you think the test code could be made more self-evident, I am open to suggestions.

**Comment by csillag** [comment]
Any further comments on this one?

File: tests/js/plugin/enhancedanchoring-test.coffee [hunk 3]
```diff
@@ -223,14 +228,13 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
       assert.include hls, anchor2.highlight[anchor2.startPage]
       assert.notInclude hls, anchor3.highlight[anchor3.startPage]
 
-  describe 'Anchor.scrollIntoView()', ->
+  describe 'Anchor.scrollToView()', ->
     it 'calls scrollIntoView() on the highlight', ->
```
Comment (tilgovi) [inline tests/js/plugin/enhancedanchoring-test.coffee]: This isn't always the case, though, right? It should call it if it's real, otherwise it scrolls to the adjacent page and then the scrolling to the highlight happens asynchronously, correct?

Comment (tilgovi) [inline tests/js/plugin/enhancedanchoring-test.coffee]: I see this is tested below. Not sure if you should make explicit here that we're _not_ using two phases.

Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > This isn't always the case, though, right? It should call it if it's real, otherwise it scrolls to the adjacent page and then the scrolling to the highlight happens asynchronously, correct?

Right. This whole section ([line 35-238](https://github.com/hypothesis/h/blob/add-highlight-test-and-change-scrolling/tests/js/plugin/enhancedanchoring-test.coffee#L35-L238)) is for testing simple, one-phase anchoring (and scrolling). The two-phase tests are [line 239-698](https://github.com/hypothesis/h/blob/add-highlight-test-and-change-scrolling/tests/js/plugin/enhancedanchoring-test.coffee#L239-L698).

Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > I see this is tested below. Not sure if you should make explicit here that we're _not_ using two phases.

Well, I suppose I could add a `describe "single-phase anchoring"` block around the single-phase tests...

Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > > I see this is tested below. Not sure if you should make explicit here that we're _not_ using two phases.

> Well, I suppose I could add a `describe "single-phase anchoring"` block around the single-phase tests...

I just did that.

**Comment by tilgovi** [comment]
I think I need to take a step back from this and maybe you should as well for a few days. We should talk about these APIs. I feel like the tests are awkward because the APIs are awkward, so it's hard for me to evaluate this. If there are not behaviors changing in this PR I'm hesitant to bother with these tests because I see things I would very much like to change, not just in implementation details but in the responsibilities and couplings between the manager, anchor, and highlights, changes that the tests would not survive.

I'll write some more thoughts soon, but I'm just going to leave it alone for a few days and enjoy the weekend.

If you and @nickstenning decide you want to merge this in the meantime, I'm not going to complain, but I see cause for some serious refactoring that we may or may not want to do now.

**Comment by nickstenning** [comment]
> I feel like the tests are awkward because the APIs are awkward, so it's hard for me to evaluate this.

This is almost certainly true, and is for me the single most useful side-effect of writing tests: finding dodgy APIs.

That said, there can be value in adding (possibly-awkward) tests to legacy code, in that those tests can serve, at least partially, as documentation of the intended use of the system under test, and can aid in refactoring that code later. (See https://amzn.com/0131177052 for more on this)

I'll take a look over this pull request shortly, although my inclination will probably be towards merging, as in my opinion having awkward code with awkward tests is usually a better situation to be in than just having awkward code.

File: tests/js/plugin/enhancedanchoring-test.coffee [hunk 4]
```diff
@@ -59,178 +36,212 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
...
-      anchor.scrollIntoView().then ->
-
-        assert.called anchor.highlight[anchor.startPage].scrollIntoView
+        assert.ok hl
+        assert.equal hl.page, page
+
+      it 'announces the creation of the highlights in an event', ->
+        am = createAnchoringManager()
+
+        assert.notCalled am.annotator.publish
```
Comment (nickstenning) [inline tests/js/plugin/enhancedanchoring-test.coffee]: I wouldn't do negative testing like this. It will cause this test to fail unnecessarily if for whatever reason you later add an event publication when the anchoring manager is created.

If you really need to ensure that the spy is only called on `createAnchor`, then reset it immediately before the call:

```javascript
am.annotator.publish.reset()
am.createAnchor(...)
assert.calledWith(am.annotator.publish, ...)
```

File: tests/js/plugin/enhancedanchoring-test.coffee [hunk 5]
```diff
@@ -59,178 +36,212 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
...
+        anchor = am.createAnchor(ann, ann.target[0]).result
+
+        hl = anchor.highlight[anchor.startPage]
+
+        assert.calledWith am.annotator.publish, 'highlightsCreated', [hl]
+
+      it 'adds an anchor property to the created Highlights', ->
+        am = createAnchoringManager()
+        ann = createTestAnnotation "a1"
+        anchor = am.createAnchor(ann, ann.target[0]).result
```
Comment (nickstenning) [inline tests/js/plugin/enhancedanchoring-test.coffee]: The same three lines appear in nearly every test in this `describe` block. This is what `before` is for -- you should use it :).

Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: > The same three lines appear in nearly every test in this `describe` block.
> This is what `before` is for -- you should use it :).

That's exactly how it was before @aron has explicitly asked me to change it so that those three lines are repeated in every single case, so that it's easier to understand what's happening.

I asked about the repetition, but he said he considers it OK.

I am find with either way, anyway.

File: tests/js/plugin/enhancedanchoring-test.coffee (hunk 4 above)
Comment (csillag) [inline tests/js/plugin/enhancedanchoring-test.coffee]: OK, I'll fix that.

File: tests/js/plugin/enhancedanchoring-test.coffee (hunk 5 above)
Comment (nickstenning) [inline tests/js/plugin/enhancedanchoring-test.coffee]: Aron said:

> My general rule of thumb is to repeat setup in the method if its just a few lines, otherwise look at ways of moving it into a setup helper. The more contained the test the better, so **it's a matter of balance**.

Which I would agree with. Here, on balance, I'd say that repeating the same three lines eight times is probably too verbose. Something like the below would be easier to read:

```coffeescript
describe 'Annotator.Plugin.EnhancedAnchoring', ->
  ...

  describe "single-phase anchoring", ->
    am = null
    ann = null
    anchor = null

    before ->
      am = createAnchoringManager()
      ann = createTestAnnotation "a1"
      anchor = am.createAnchor(ann, ann.target[0]).result

    it 'creates the anchors from the right targets', ->
      assert.equal anchor.id, "fake anchor for " + anchor.target.id

    it '...', ->
      ...
```

**Comment by nickstenning** [comment]
Good stuff.

For what it's worth, I agree with @tilgovi's assessment -- the code being tested here is hard to follow and we definitely need to look at it again, preferably soon. Two-phase anchoring in particular is complicated, and (for example) the asynchronous nature of `Anchor#scrollToView()` is not at all obvious to me.

That said, I do believe it's better to have it tested than not, as this:

- exposes the fact that the API could do with improvement (because the tests are hard to write)
- makes it easier to make incremental (rather than big-bang) improvements to this part of the code
- limits the extent to which unrelated changes can break this part of the code
- serves as secondary documentation on how this API is (currently) supposed to work

So I'm merging it.
